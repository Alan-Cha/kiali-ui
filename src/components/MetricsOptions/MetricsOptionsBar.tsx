import * as React from 'react';
import { Toolbar, ToolbarRightContent, FormGroup } from 'patternfly-react';
import isEqual from 'lodash/fp/isEqual';

import Refresh from '../Refresh/Refresh';
import { ToolbarDropdown } from '../ToolbarDropdown/ToolbarDropdown';
import history, { URLParams, HistoryManager } from '../../app/History';
import { config } from '../../config';
import MetricsOptions from '../../types/MetricsOptions';
import { MetricsDirection } from '../../types/Metrics';

import { MetricsSettings, Quantiles, MetricsSettingsDropdown } from './MetricsSettings';
import { MetricsLabels as L } from './MetricsLabels';

interface Props {
  onOptionsChanged: (opts: MetricsOptions) => void;
  onReporterChanged: (reporter: string) => void;
  onRefresh: () => void;
  onLabelsFiltersChanged: (label: L.LabelName, value: string, checked: boolean) => void;
  metricReporter: string;
  direction: MetricsDirection;
  labelValues: Map<L.LabelName, L.LabelValues>;
}

export class MetricsOptionsBar extends React.Component<Props> {
  static PollIntervals = config().toolbar.pollInterval;
  static DefaultPollInterval = config().toolbar.defaultPollInterval;

  static Durations = config().toolbar.intervalDuration;
  static DefaultDuration = config().toolbar.defaultDuration;

  static ReporterOptions: { [key: string]: string } = {
    destination: 'Destination',
    source: 'Source'
  };

  private metricsSettings: MetricsSettings;
  private duration: number;
  private pollInterval: number;
  private shouldReportOptions: boolean;

  static initialPollInterval = (urlParams: URLSearchParams): number => {
    const pi = urlParams.get(URLParams.POLL_INTERVAL);
    if (pi !== null) {
      return Number(pi);
    }
    return MetricsOptionsBar.DefaultPollInterval;
  };

  static initialDuration = (urlParams: URLSearchParams): number => {
    let d = urlParams.get(URLParams.DURATION);
    if (d !== null) {
      sessionStorage.setItem(URLParams.DURATION, d);
      return Number(d);
    }
    d = sessionStorage.getItem(URLParams.DURATION);
    return d !== null ? Number(d) : MetricsOptionsBar.DefaultDuration;
  };

  static initialMetricsSettings = (urlParams: URLSearchParams): MetricsSettings => {
    const settings: MetricsSettings = {
      showAverage: true,
      showQuantiles: ['0.5', '0.95', '0.99'],
      activeLabels: []
    };
    const avg = urlParams.get(URLParams.SHOW_AVERAGE);
    if (avg !== null) {
      settings.showAverage = avg === 'true';
    }
    const quantiles = urlParams.get(URLParams.QUANTILES);
    if (quantiles !== null) {
      if (quantiles.trim().length !== 0) {
        settings.showQuantiles = quantiles.split(' ').map(val => val.trim() as Quantiles);
      } else {
        settings.showQuantiles = [];
      }
    }
    const byLabels = urlParams.getAll(URLParams.BY_LABELS);
    if (byLabels.length !== 0) {
      settings.activeLabels = byLabels as L.LabelName[];
    }
    return settings;
  };

  constructor(props: Props) {
    super(props);
  }

  componentDidMount() {
    // Init state upstream
    this.reportOptions();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.shouldReportOptions) {
      this.shouldReportOptions = false;
      this.reportOptions();
    }
  }

  onPollIntervalChanged = (key: number) => {
    HistoryManager.setParam(URLParams.POLL_INTERVAL, String(key));
  };

  onDurationChanged = (key: number) => {
    sessionStorage.setItem(URLParams.DURATION, String(key));
    HistoryManager.setParam(URLParams.DURATION, String(key));
  };

  reportOptions() {
    // State-to-options converter (removes unnecessary properties)
    let labelsIn: L.PromLabel[] = [];
    let labelsOut: L.PromLabel[] = [];
    if (this.props.direction === MetricsDirection.INBOUND) {
      labelsIn = this.metricsSettings.activeLabels.map(lbl => L.INBOUND_LABELS.get(lbl)!);
    } else {
      labelsOut = this.metricsSettings.activeLabels.map(lbl => L.OUTBOUND_LABELS.get(lbl)!);
    }
    this.props.onOptionsChanged({
      duration: this.duration,
      byLabelsIn: labelsIn,
      byLabelsOut: labelsOut,
      avg: this.metricsSettings.showAverage,
      quantiles: this.metricsSettings.showQuantiles
    });
  }

  onReporterChanged = (reporter: string) => {
    HistoryManager.setParam(URLParams.REPORTER, reporter);
    this.props.onReporterChanged(reporter);
  };

  onMetricsSettingsChanged = (settings: MetricsSettings) => {
    const urlParams = new URLSearchParams(history.location.search);
    urlParams.set(URLParams.SHOW_AVERAGE, String(settings.showAverage));
    urlParams.set(URLParams.QUANTILES, settings.showQuantiles.join(' '));
    urlParams.delete(URLParams.BY_LABELS);
    settings.activeLabels.forEach(lbl => urlParams.append(URLParams.BY_LABELS, lbl));
    history.replace(history.location.pathname + '?' + urlParams.toString());
  };

  render() {
    this.processUrlParams();

    return (
      <Toolbar>
        <FormGroup>
          <MetricsSettingsDropdown
            onChanged={this.onMetricsSettingsChanged}
            onLabelsFiltersChanged={this.props.onLabelsFiltersChanged}
            labelValues={this.props.labelValues}
            {...this.metricsSettings}
          />
        </FormGroup>
        <FormGroup>
          <ToolbarDropdown
            id={'metrics_filter_reporter'}
            disabled={false}
            handleSelect={this.onReporterChanged}
            nameDropdown={'Reported from'}
            value={this.props.metricReporter}
            initialLabel={MetricsOptionsBar.ReporterOptions[this.props.metricReporter]}
            options={MetricsOptionsBar.ReporterOptions}
          />
        </FormGroup>
        <ToolbarDropdown
          id={'metrics_filter_interval_duration'}
          disabled={false}
          handleSelect={this.onDurationChanged}
          nameDropdown={'Displaying'}
          initialValue={this.duration}
          initialLabel={String(MetricsOptionsBar.Durations[this.duration])}
          options={MetricsOptionsBar.Durations}
        />
        <ToolbarRightContent>
          <Refresh
            id="metrics-refresh"
            handleRefresh={this.props.onRefresh}
            onSelect={this.onPollIntervalChanged}
            pollInterval={this.pollInterval}
          />
        </ToolbarRightContent>
      </Toolbar>
    );
  }

  private processUrlParams = () => {
    const urlParams = new URLSearchParams(history.location.search);
    const pollInterval = MetricsOptionsBar.initialPollInterval(urlParams);
    const duration = MetricsOptionsBar.initialDuration(urlParams);
    const metricsSettings = MetricsOptionsBar.initialMetricsSettings(urlParams);

    this.shouldReportOptions =
      pollInterval !== this.pollInterval ||
      duration !== this.duration ||
      !isEqual(metricsSettings)(this.metricsSettings);

    this.pollInterval = pollInterval;
    this.duration = duration;
    this.metricsSettings = metricsSettings;
  };
}

export default MetricsOptionsBar;
