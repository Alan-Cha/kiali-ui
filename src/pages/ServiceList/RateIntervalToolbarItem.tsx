import * as React from 'react';
import MetricsOptionsBar from '../../components/MetricsOptions/MetricsOptionsBar';
import { DropdownButton, MenuItem } from 'patternfly-react';

type RateIntervalToolbarItemProps = {
  rateIntervalSelected: string;
  onRateIntervalChanged?: (key: string) => void;
};

export default class RateIntervalToolbarItem extends React.Component<RateIntervalToolbarItemProps> {
  render() {
    const rateIntervalSelected = MetricsOptionsBar.RateIntervals.find(el => {
      return el[0] === this.props.rateIntervalSelected;
    });

    return (
      <div className="form-group">
        <label style={{ paddingRight: '0.5em' }}>Rate Interval:</label>
        <DropdownButton
          title={'Last ' + rateIntervalSelected![1]}
          onSelect={this.props.onRateIntervalChanged}
          id="rateIntervalDropDown"
        >
          {MetricsOptionsBar.RateIntervals.map(r => (
            <MenuItem key={r[0]} active={r[0] === this.props.rateIntervalSelected} eventKey={r[0]}>
              Last {r[1]}
            </MenuItem>
          ))}
        </DropdownButton>
      </div>
    );
  }
}
