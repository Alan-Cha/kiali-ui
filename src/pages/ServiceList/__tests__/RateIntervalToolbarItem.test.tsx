import * as React from 'react';
import { shallow } from 'enzyme';
import RateIntervalToolbarItem from '../RateIntervalToolbarItem';
import MetricsOptionsBar from '../../../components/MetricsOptions/MetricsOptionsBar';

describe('RateIntervalToolbarItem', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<RateIntervalToolbarItem rateIntervalSelected="10m" />);
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('MenuItem').length).toBe(MetricsOptionsBar.RateIntervals.length);
  });
});
