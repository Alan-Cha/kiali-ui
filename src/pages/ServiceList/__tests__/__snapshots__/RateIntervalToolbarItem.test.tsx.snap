// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateIntervalToolbarItem renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RateIntervalToolbarItem
    rateIntervalSelected="10m"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          style={
            Object {
              "paddingRight": "0.5em",
            }
          }
        >
          Rate Interval:
        </label>,
        <DropdownButton
          id="rateIntervalDropDown"
          onSelect={undefined}
          title="Last 10 minutes"
        >
          <MenuItem
            active={false}
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="1m"
            header={false}
          >
            Last 
            1 minute
          </MenuItem>
          <MenuItem
            active={false}
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="5m"
            header={false}
          >
            Last 
            5 minutes
          </MenuItem>
          <MenuItem
            active={true}
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="10m"
            header={false}
          >
            Last 
            10 minutes
          </MenuItem>
          <MenuItem
            active={false}
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="30m"
            header={false}
          >
            Last 
            30 minutes
          </MenuItem>
        </DropdownButton>,
      ],
      "className": "form-group",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Rate Interval:",
          "style": Object {
            "paddingRight": "0.5em",
          },
        },
        "ref": null,
        "rendered": "Rate Interval:",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <MenuItem
              active={false}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="1m"
              header={false}
            >
              Last 
              1 minute
            </MenuItem>,
            <MenuItem
              active={false}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="5m"
              header={false}
            >
              Last 
              5 minutes
            </MenuItem>,
            <MenuItem
              active={true}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="10m"
              header={false}
            >
              Last 
              10 minutes
            </MenuItem>,
            <MenuItem
              active={false}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="30m"
              header={false}
            >
              Last 
              30 minutes
            </MenuItem>,
          ],
          "id": "rateIntervalDropDown",
          "onSelect": undefined,
          "title": "Last 10 minutes",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1m",
            "nodeType": "class",
            "props": Object {
              "active": false,
              "bsClass": "dropdown",
              "children": Array [
                "Last ",
                "1 minute",
              ],
              "disabled": false,
              "divider": false,
              "eventKey": "1m",
              "header": false,
            },
            "ref": null,
            "rendered": Array [
              "Last ",
              "1 minute",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "5m",
            "nodeType": "class",
            "props": Object {
              "active": false,
              "bsClass": "dropdown",
              "children": Array [
                "Last ",
                "5 minutes",
              ],
              "disabled": false,
              "divider": false,
              "eventKey": "5m",
              "header": false,
            },
            "ref": null,
            "rendered": Array [
              "Last ",
              "5 minutes",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "10m",
            "nodeType": "class",
            "props": Object {
              "active": true,
              "bsClass": "dropdown",
              "children": Array [
                "Last ",
                "10 minutes",
              ],
              "disabled": false,
              "divider": false,
              "eventKey": "10m",
              "header": false,
            },
            "ref": null,
            "rendered": Array [
              "Last ",
              "10 minutes",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "30m",
            "nodeType": "class",
            "props": Object {
              "active": false,
              "bsClass": "dropdown",
              "children": Array [
                "Last ",
                "30 minutes",
              ],
              "disabled": false,
              "divider": false,
              "eventKey": "30m",
              "header": false,
            },
            "ref": null,
            "rendered": Array [
              "Last ",
              "30 minutes",
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            style={
              Object {
                "paddingRight": "0.5em",
              }
            }
          >
            Rate Interval:
          </label>,
          <DropdownButton
            id="rateIntervalDropDown"
            onSelect={undefined}
            title="Last 10 minutes"
          >
            <MenuItem
              active={false}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="1m"
              header={false}
            >
              Last 
              1 minute
            </MenuItem>
            <MenuItem
              active={false}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="5m"
              header={false}
            >
              Last 
              5 minutes
            </MenuItem>
            <MenuItem
              active={true}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="10m"
              header={false}
            >
              Last 
              10 minutes
            </MenuItem>
            <MenuItem
              active={false}
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="30m"
              header={false}
            >
              Last 
              30 minutes
            </MenuItem>
          </DropdownButton>,
        ],
        "className": "form-group",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Rate Interval:",
            "style": Object {
              "paddingRight": "0.5em",
            },
          },
          "ref": null,
          "rendered": "Rate Interval:",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <MenuItem
                active={false}
                bsClass="dropdown"
                disabled={false}
                divider={false}
                eventKey="1m"
                header={false}
              >
                Last 
                1 minute
              </MenuItem>,
              <MenuItem
                active={false}
                bsClass="dropdown"
                disabled={false}
                divider={false}
                eventKey="5m"
                header={false}
              >
                Last 
                5 minutes
              </MenuItem>,
              <MenuItem
                active={true}
                bsClass="dropdown"
                disabled={false}
                divider={false}
                eventKey="10m"
                header={false}
              >
                Last 
                10 minutes
              </MenuItem>,
              <MenuItem
                active={false}
                bsClass="dropdown"
                disabled={false}
                divider={false}
                eventKey="30m"
                header={false}
              >
                Last 
                30 minutes
              </MenuItem>,
            ],
            "id": "rateIntervalDropDown",
            "onSelect": undefined,
            "title": "Last 10 minutes",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1m",
              "nodeType": "class",
              "props": Object {
                "active": false,
                "bsClass": "dropdown",
                "children": Array [
                  "Last ",
                  "1 minute",
                ],
                "disabled": false,
                "divider": false,
                "eventKey": "1m",
                "header": false,
              },
              "ref": null,
              "rendered": Array [
                "Last ",
                "1 minute",
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "5m",
              "nodeType": "class",
              "props": Object {
                "active": false,
                "bsClass": "dropdown",
                "children": Array [
                  "Last ",
                  "5 minutes",
                ],
                "disabled": false,
                "divider": false,
                "eventKey": "5m",
                "header": false,
              },
              "ref": null,
              "rendered": Array [
                "Last ",
                "5 minutes",
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "10m",
              "nodeType": "class",
              "props": Object {
                "active": true,
                "bsClass": "dropdown",
                "children": Array [
                  "Last ",
                  "10 minutes",
                ],
                "disabled": false,
                "divider": false,
                "eventKey": "10m",
                "header": false,
              },
              "ref": null,
              "rendered": Array [
                "Last ",
                "10 minutes",
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "30m",
              "nodeType": "class",
              "props": Object {
                "active": false,
                "bsClass": "dropdown",
                "children": Array [
                  "Last ",
                  "30 minutes",
                ],
                "disabled": false,
                "divider": false,
                "eventKey": "30m",
                "header": false,
              },
              "ref": null,
              "rendered": Array [
                "Last ",
                "30 minutes",
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
